refactor(animations): Add loading strategy for the Async Animations

JeanMeche opened the pull request:
In some cases apps need to schedule a bit later the loading of the animation module. This change add a new strategy based of `afterNextRender`

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JeanMeche commented on packages/platform-browser/animations/async/src/providers.ts:
```diff
@@ -48,13 +48,14 @@ import {AsyncAnimationRendererFactory} from './async_animation_renderer';
  */
 export function provideAnimationsAsync(
   type: 'animations' | 'noop' = 'animations',
+  loadingStrategy: 'default' | 'afterNextRender' = 'default',
```
Probably needs some bikeschedding as we also want to rename `after*Render`.

AndrewKushnir commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -68,9 +70,22 @@ export class AsyncAnimationRendererFactory implements OnDestroy, RendererFactory
+      ɵcreateEngine: (type: 'animations' | 'noop', doc: Document) => AnimationEngine;
+      ɵAnimationRendererFactory: typeof AnimationRendererFactory;
+    }>;
+    if (this.loadingStrategy === 'default') {
+      moduleImplPromise = this.moduleImpl ?? import('@angular/animations/browser').then((m) => m);
+    } else {
+      moduleImplPromise = new Promise((res) => {
+        afterNextRender(() => {
+          res(this.moduleImpl ?? import('@angular/animations/browser').then((m) => m));
+        });
+      });
```
```suggestion
    const loadingSchedulerFn = this.injector.get(ɵASYNC_ANIMATION_LOADING_SCHEDULER_FN, {optional: true});
    if (loadingSchedulerFn) {
      moduleImplPromise = loadingSchedulerFn(this.moduleImpl ?? import('@angular/animations/browser').then((m) => m));
    } else {
      moduleImplPromise = this.moduleImpl ?? import('@angular/animations/browser').then((m) => m);
    }
```

JeanMeche replied:
Sounds a very reasonable idea !

AndrewKushnir reviewed (commented):
I'm wondering if we should make a private token for now, which would return a function that contains scheduling logic (see comment below). That'd give us some flexibility to experiment with exact timing.

@jelbourn WDYT?

JeanMeche committed:
refactor(animations): Add loading strategy for the Async Animations

In some cases apps need to schedule a bit later the loading of the animation module. This private token will allow to investigate which other strategy could be useful.

alxhub reviewed (approved):
This should be a `refactor` commit as it does not implement a user-visible feature.

JeanMeche commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -68,9 +71,23 @@ export class AsyncAnimationRendererFactory implements OnDestroy, RendererFactory
+      ɵcreateEngine: (type: 'animations' | 'noop', doc: Document) => AnimationEngine;
+      ɵAnimationRendererFactory: typeof AnimationRendererFactory;
+    }>;
+    const loadingSchedulerFn = this.injector.get(ɵASYNC_ANIMATION_LOADING_SCHEDULER_FN, undefined, {
```
Makes sense.

jelbourn commented on packages/platform-browser/animations/async/src/async_animation_renderer.ts:
```diff
@@ -68,9 +73,20 @@ export class AsyncAnimationRendererFactory implements OnDestroy, RendererFactory
+      ɵAnimationRendererFactory: typeof AnimationRendererFactory;
+    }>;
+
+    const loadFn = () => this.moduleImpl ?? import('@angular/animations/browser').then((m) => m);
```
Does this `then` do anything?

JeanMeche replied:
Yes, the related comment got misplaced.

jelbourn reviewed (approved):
LGTM

AndrewKushnir commented:
[Presubmit](http://test/OCL:667720925:BASE:667720965:1724708425178:c57b931c).

alxhub commented:
This PR was merged into the repository by commit d73f2e91a22e77f5879e164c40a5d2eb6fb70333.

The changes were merged into the following branches: main, 18.2.x
